### PR TITLE
refactor: use native validation for parameter form

### PR DESCRIPTION
**Pull request by TrySound**
Here switched to native html solution for validation in parameter variable form. Added three new form utilities to replicate existing behavior with closing errored form only on the second attempt when all errors are shown.

- Form
- useFormField
- checkCanRequestSubmit

These are a few important apis for validation
- form.requestSubmit()
- form.checkValidity()
- input.checkValidity()
- input.setCustomValidity('message')

image

### CODE REVIEW

File: apps/builder/app/shared/form-utils/form-utils.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    allErrorsVisible: fields.every((field) => field.isErrorVisible),
+    showAllErrors: () => {
+      for (const field of fields) {
+        field.onBlur();
+      }
+    },
+  };
+};
+
+export const useFormField = ({
```
[the remark this answers is not in the document]
Comment (TrySound) [reply apps/builder/app/shared/form-utils/form-utils.tsx:70]: useFormField relies on native validation,
useField is old implementation, git did not diffed renaming well

[the remark this answers is not in the document]
Comment (kof) [reply apps/builder/app/shared/form-utils/form-utils.tsx:76]: then we need to call it useFieldDeprecated?

[the remark this answers is not in the document]
Comment (kof) [reply apps/builder/app/shared/form-utils/form-utils.tsx:76]: just so when we touch deprecated code, we migrate it to the new one

[the remark this answers is not in the document]
Comment (kof) [reply apps/builder/app/shared/form-utils/form-utils.tsx:76]: I saw the @deprecated thing, but I think its not visible enough to really incentivice to refactor

[the remark this answers is not in the document]
Comment (TrySound) [reply apps/builder/app/shared/form-utils/form-utils.tsx:76]: visible enough, we use it only in variable popover which I'm gonna rebuild anyway

**Comment by kof** [comment]
- [ ] Looks like its not saving on field blur

image

**Comment by TrySound** [comment]
It never saved on blur

**Comment by kof** [comment]
- [ ] when changing the field then hitting ESC, do you think we should just revert the change, then on second ESC close the panel? Right now ESC acts as "save".

**Comment by kof** [comment]
>  It never saved on blur

Lets get the desired behavior here? without waiting for a second PR?

**Comment by TrySound** [comment]
It never saved on blur because we use the same forms for "create" and "edit". Form will loose focus if save on blur.

**Comment by TrySound** [comment]
>  when changing the field then hitting ESC, do you think we should just revert the change, then on second ESC close the panel? Right now ESC acts as "save".

What change to revert if you want to save on blur?

**Comment by kof** [comment]
> It never saved on blur because we use the same forms for "create" and "edit". Form will loose focus if save on blur.

Maybe we need to differentiate the 2 cases? Because it feels much better when updating stuff when it applies right away and indeed when creating new stuff, you don't want it to create before you actively said you are done.

**Comment by kof** [comment]
> What change to revert if you want to save on blur?

Its tricky. When I changed something and it didn't save right away - then I am probably expecting it with ESC to not save it at all. When I changed it but it was visibly auto-saved (I can see it saved), then ESC is probably just expected to close the thing, wdyt? We can compare with some other UX best practices. This is my hunch

**Comment by kof** [comment]
My biggest beef with our current behavior in some cases is that when I change a value and it is represented somewhere else - I don't see it being updated, I feel like if I close it, it will be gone, there is no clarity what happens if I close it now.

When everything is auto-saved all the time and reflected immediately everywhere - I feel most safe, but  there are cases like you mentioned where I am creating a new item and don't expect it to be created until I do something confirmational like hitting ENTER

**Comment by kof** [comment]
Did some testing with figma on behavior https://share.descript.com/view/8ETeg4IxS21

**Review by istarkov — APPROVED** [review]

**Comment by kof** [comment]
Is this still needed?

**Pull request closed without merging**